core/src/main/java/org/apache/iceberg/BaseTableScan.java

```diff
@@ -86,20 +90,56 @@ public TableScan appendsAfter(long fromSnapshotId) {
   public TableScan useSnapshot(long scanSnapshotId) {
     Preconditions.checkArgument(
         snapshotId() == null, "Cannot override snapshot, already set to id=%s", snapshotId());
+    Preconditions.checkArgument(
+        ref() == null, "Cannot override snapshot, already set to id=%s", ref());
     Preconditions.checkArgument(
         tableOps().current().snapshot(scanSnapshotId) != null,
         "Cannot find snapshot with ID %s",
         scanSnapshotId);
     return newRefinedScan(
-        tableOps(), table(), tableSchema(), context().useSnapshotId(scanSnapshotId));
+        tableOps(),
+        table(),
+        tableSchema(),
+        context().useRef(SnapshotRef.MAIN_BRANCH).useSnapshotId(scanSnapshotId));
```

Comment:
Why does this set `useRef`? The precondition above checks that `ref` is `null`, so it seems like it should never be set if the snapshot ID is set.

Reply:
For the scan from ref cases, we set both a ref and snapshot so setting it here to be consistent with that. It maintains the invariant that every scan context will have a ref and a snapshot.

 As for the null ref check, I think we can remove that because if useRef is called, we'll be setting both, and then the existing snapshot check will fail.